src/io/rkn/conformity.clj

```diff
@@ -219,41 +212,61 @@
                      :where [?tx ?na ?nv]]
                    db conformity-attr norm)))))
 
-(defn handle-first-time-only-norm
-  [acc conn norm-attr norm-map norm-name sync-schema-timeout tx-instant]
-  (if (first-time-only-conforms-to? (db conn) norm-attr norm-name)
-    acc
-    (let [{:keys [txes ex]} (get-norm conn norm-map norm-name)]
-      (handle-txes acc conn norm-attr norm-name txes ex sync-schema-timeout
-        tx-instant))))
-
-(defn handle-mutable-norm
-  [acc conn norm-attr norm-map norm-name sync-schema-timeout tx-instant]
-  (let [{:keys [txes ex]} (get-norm conn norm-map norm-name)]
-    (if (conforms-to? (db conn) norm-attr norm-name (count txes))
-      acc
-      (handle-txes acc conn norm-attr norm-name txes ex sync-schema-timeout
-        tx-instant))))
-
-(defn reduce-norms
+(defn- handle-first-time-only-norm
+  [conn norm-attr norm-map norm-name sync-schema-timeout tx-instant]
+  (when-not (first-time-only-conforms-to? (db conn) norm-attr norm-name)
+    (let [{:keys [txes]} (get-norm conn norm-map norm-name)]
+      (handle-txes conn norm-attr norm-name txes sync-schema-timeout
+                   tx-instant))))
+
+(defn- handle-mutable-norm
+  [conn norm-attr norm-map norm-name sync-schema-timeout tx-instant]
+  (let [{:keys [txes]} (get-norm conn norm-map norm-name)]
+    (when-not (conforms-to? (db conn) norm-attr norm-name (count txes))
+      (handle-txes conn norm-attr norm-name txes sync-schema-timeout
+                   tx-instant))))
+
+(defn- reduce-norms*
+  [applied-norms conn norm-attr norm-map norm-names tx-instant]
+  (let [sync-schema-timeout (:conformity.setting/sync-schema-timeout norm-map)]
+    (reduce
+     (fn [{:as acc :keys [applied-norms]} norm-name]
+       (let [{:keys [requires first-time-only]} (get norm-map norm-name)
+             unapplied-requires (remove applied-norms requires)
+             requires-results (reduce-norms* applied-norms conn norm-attr norm-map
+                                            unapplied-requires
+                                            tx-instant)
+             acc (-> acc
+                     (update :results concat (:results requires-results))
+                     (update :applied-norms into (:applied-norms requires-results)))]
+         (if (contains? (:applied-norms acc) norm-name)
+           acc
+           (try
+             (-> acc
+                 (update :results concat
+                         (if first-time-only
+                           (handle-first-time-only-norm conn norm-attr norm-map norm-name
+                                                        sync-schema-timeout tx-instant)
+                           (handle-mutable-norm conn norm-attr norm-map norm-name
+                                                sync-schema-timeout tx-instant)))
+                 (update :applied-norms conj norm-name))
```

Comment:
This is the heart of the change; we keep a set of norms we have processed in the current session, and skip processing for a norm if we see it again in the dependency tree.

It may be better to topologically sort the dependency tree and walk it once, but that would have been a bigger change.